flake.nix

```diff
@@ -0,0 +1,254 @@
+{
+  description = "Emacs versions for CI";
+
+  inputs = {
+    nixpkgs.url = "nixpkgs/nixpkgs-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
+
+    flake-compat = { url = "github:edolstra/flake-compat"; flake = false; };
+    "emacs-23.4" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-23.4.tar.gz"; flake = false; };
+    "emacs-24.1" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-24.1.tar.gz"; flake = false; };
+    "emacs-24.2" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-24.2.tar.gz"; flake = false; };
+    "emacs-24.3" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-24.3.tar.gz"; flake = false; };
+    "emacs-24.4" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-24.4.tar.gz"; flake = false; };
+    "emacs-24.5" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-24.5.tar.gz"; flake = false; };
+    "emacs-25.1" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-25.1.tar.gz"; flake = false; };
+    "emacs-25.2" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-25.2.tar.gz"; flake = false; };
+    "emacs-25.3" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-25.3.tar.gz"; flake = false; };
+    "emacs-26.1" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-26.1.tar.gz"; flake = false; };
+    "emacs-26.2" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-26.2.tar.gz"; flake = false; };
+    "emacs-26.3" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-26.3.tar.gz"; flake = false; };
+    "emacs-27.1" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-27.1.tar.gz"; flake = false; };
+    "emacs-27.2" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-27.2.tar.gz"; flake = false; };
+    "emacs-28.1" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-28.1.tar.gz"; flake = false; };
+    "emacs-28.2" = { url = "https://ftp.gnu.org/gnu/emacs/emacs-28.2.tar.gz"; flake = false; };
+    emacs-snapshot = { url = "github:emacs-mirror/emacs"; flake = false; };
+    emacs-release-snapshot = { url = "github:emacs-mirror/emacs?ref=emacs-29"; flake = false; };
+    macos-unexec-patches = { url = "https://github.com/emacs-mirror/emacs/commit/888ffd960c06d56a409a7ff15b1d930d25c56089.patch"; flake = false; };
+    latest-package-keyring = { url = "github:emacs-mirror/emacs?path=/etc/package-keyring.gpg"; flake = false; };
```

Comment:
Unless you need to pin this to a separate revision, you can take the `emacs-snapshot` input:

```nix
"${inputs.emacs-snapshot}/etc/package-keyring.gpg"
```

Reply:
Yeah, but I wouldn't want lots of builds to depend on the full tarball for the latest build.

Reply:
An alternative would be to use a filtering library like [nix-filter](https://github.com/numtide/nix-filter). And it's possible to have separate lock entries pointing to the same upstream URL, so you don't necessary have to use `dir` parameter.